include/lug/System/Exception.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <string>
+#include <lug/System/Export.hpp>
+
+namespace lug {
+namespace System {
+
+class LUG_SYSTEM_API Exception {
+public:
+    Exception(const char *typeName, const std::string &description, const char* file, const char* function, long line);
+    Exception(Exception&& other);
+    Exception(const Exception& other);
+    Exception& operator=(const Exception& other);
+    Exception& operator=(const Exception&& other);
+    ~Exception() throw() {}
+
+    const std::string&              getTypeName() const;
+    const std::string&              getDescription() const;
+    const std::string&              getFile() const;
+    const std::string&              getFunction() const;
+    const long                      getLine() const;
+
+    const char*                     what() const;
+
+private:
+    std::string                     _typeName;
+    std::string                     _description;
+    std::string                     _file;
+    std::string                     _function;
+    long                            _line;
+
+    mutable std::string             _fullDesc;
+};
+
+
+class NotImplementedException: public Exception {
+public:
+    NotImplementedException(const std::string &description, const char* file, const char* function, long line)
+    : Exception{"NotImplementedException", description, file, function, line} {}
+};
+
+
+class FileNotFoundException: public Exception {
+public:
+    FileNotFoundException(const std::string &description, const char* file, const char* function, long line)
+    : Exception{"FileNotFoundException", description, file, function, line} {}
+};
+
+
+class IOException: public Exception {
+public:
+    IOException(const std::string &description, const char* file, const char* function, long line)
+    : Exception{"IOException", description, file, function, line} {}
+};
+
+
+class InvalidParametersException: public Exception {
+public:
+    InvalidParametersException(const std::string &description, const char* file, const char* function, long line)
+    : Exception{"InvalidParametersException", description, file, function, line} {}
+};
+
+
+class InternalErrorException: public Exception {
+public:
+    InternalErrorException(const std::string &description, const char* file, const char* function, long line)
+    : Exception{"InternalErrorException", description, file, function, line} {}
+};
+
+
+class RendererException: public Exception {
+public:
+    RendererException(const std::string &description, const char* file, const char* function, long line)
+    : Exception{"RendererException", description, file, function, line} {}
+};
+
+
+class WindowException: public Exception {
+public:
+    WindowException(const std::string &description, const char* file, const char* function, long line)
+    : Exception{"WindowException", description, file, function, line} {}
+};
+
+
+#define LUG_EXCEPT(type, desc)\
+do {\
+  throw type(desc, __FILE__, __FUNCTION__, __LINE__);\
+} while (0)
+
+}
+}
```

Comment:
Please add a `\n` at the end of the file. Configure your editor to do it for you.

In sublime text, the required settings for the project are:
```
{
    "tab_size": 4,
    "translate_tabs_to_spaces": true,
    "trim_trailing_white_space_on_save": true,
    "ensure_newline_at_eof_on_save": true,
}
```